File: .agents/skills/query-homelab-state/scripts/query_infra.js
const { execSync } = require('child_process');

function runCmd(cmd) {
    try {
        const stdout = execSync(cmd, { encoding: 'utf8', stdio: ['pipe', 'pipe', 'pipe'] });
        return { stdout: stdout.trim(), stderr: '', code: 0 };
    } catch (err) {
        return { stdout: err.stdout ? err.stdout.toString().trim() : '', stderr: err.stderr ? err.stderr.toString().trim() : err.message, code: err.status || 1 };
    }
}

function queryDocker(service, fetchLogs) {
    const ps = runCmd(`docker ps -a --filter name=${service} --format '{{.ID}}|{{.Names}}|{{.Status}}'`);
    if (ps.code !== 0 || !ps.stdout) return { found: false };

    const lines = ps.stdout.split('\n');
    const [cid, cname, status] = lines[0].split('|');

    let report = `## 🐳 Platform: Docker\n**Service**: ${cname} (${cid})\n**Status**: ${status}\n\n`;

    const stats = runCmd(`docker stats ${cid} --no-stream --format '{{.CPUPerc}} CPU | {{.MemUsage}} RAM'`);
    if (stats.code === 0 && stats.stdout) {
        report += `**Resources**: ${stats.stdout}\n\n`;
    }

    if (fetchLogs) {
        const logs = runCmd(`docker logs --tail=100 ${cid}`);
        report += "### 📝 Logs (Last 100 lines)\n```text\n";
        const combined = `${logs.stdout}\n${logs.stderr}`.trim();
        report += (combined ? combined : '<empty>') + "\n```\n";
    }

    return { found: true, report };
}

function queryKubernetes(service, fetchLogs) {
    let podName, status;
    const json = runCmd(`kubectl get pods -l app=${service} -o json`);
    
    if (json.code === 0 && json.stdout) {
        try {
            const data = JSON.parse(json.stdout);
            if (data.items && data.items.length > 0) {
                podName = data.items[0].metadata.name;
                status = data.items[0].status.phase;
            }
        } catch (e) {}
    }

    if (!podName) {
        const fallback = runCmd(`kubectl get pods --no-headers | grep ${service} | head -n 1 | awk '{print $1"|"$3}'`);
        if (fallback.code !== 0 || !fallback.stdout) return { found: false };
        [podName, status] = fallback.stdout.split('|');
    }

    let report = `## ☸️ Platform: Kubernetes\n**Service**: ${podName}\n**Status**: ${status}\n\n`;

    const top = runCmd(`kubectl top pod ${podName} --no-headers`);
    if (top.code === 0 && top.stdout) {
        report += `**Resources (CPU/RAM)**: ${top.stdout}\n\n`;
    }

    if (fetchLogs) {
        const logs = runCmd(`kubectl logs --tail=100 ${podName}`);
        report += "### 📝 Logs (Last 100 lines)\n```text\n";
        const combined = `${logs.stdout}\n${logs.stderr}`.trim();
        report += (combined ? combined : '<empty>') + "\n```\n";
    }

    return { found: true, report };
}

function main() {
    if (process.argv.length < 3) {
        console.log('Usage: node query_infra.js <target_service> [fetch_logs: true/false]');
        process.exit(1);
    }

    const service = process.argv[2];
    const fetchLogs = process.argv.length > 3 && process.argv[3].toLowerCase() === 'true';

    console.log(`# Infrastructure Report for '${service}'\n`);

    const dockerRes = queryDocker(service, fetchLogs);
    if (dockerRes.found) {
        console.log(dockerRes.report);
        process.exit(0);
    }

    const k8sRes = queryKubernetes(service, fetchLogs);
    if (k8sRes.found) {
        console.log(k8sRes.report);
        process.exit(0);
    }

    console.log(`[ERROR] Service '${service}' not found in local Docker or Kubernetes context.`);
    process.exit(1);
}

main();

File: .agents/skills/query-homelab-state/scripts/query_infra.py
import sys
import subprocess
import json

def run_cmd(cmd):
    try:
        result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
        return result.stdout.strip(), result.stderr.strip(), result.returncode
    except Exception as e:
        return "", str(e), 1

def query_docker(service, fetch_logs):
    # Find container ID
    stdout, stderr, code = run_cmd(f"docker ps -a --filter name={service} --format '{{{{.ID}}}}|{{{{.Names}}}}|{{{{.Status}}}}'")
    if code != 0 or not stdout:
        return False, None
    
    parts = stdout.split('\n')[0].split('|')
    cid = parts[0]
    cname = parts[1]
    status = parts[2]
    
    report = f"## 🐳 Platform: Docker\n**Service**: {cname} ({cid})\n**Status**: {status}\n\n"
    
    # Get stats
    stats_out, _, stats_code = run_cmd(f"docker stats {cid} --no-stream --format '{{{{.CPUPerc}}}} CPU | {{{{.MemUsage}}}} RAM'")
    if stats_code == 0 and stats_out:
        report += f"**Resources**: {stats_out}\n\n"
        
    if fetch_logs:
        logs_out, logs_err, logs_code = run_cmd(f"docker logs --tail=100 {cid}")
        report += "### 📝 Logs (Last 100 lines)\n```text\n"
        # Combine stdout and stderr for logs
        combined_logs = (logs_out + "\n" + logs_err).strip()
        report += (combined_logs if combined_logs else "<empty>") + "\n```\n"
        
    return True, report

def query_kubernetes(service, fetch_logs):
    stdout, stderr, code = run_cmd(f"kubectl get pods -l app={service} -o json")
    if code != 0 or not stdout:
        # Fallback to grep if no labels
        stdout, stderr, code = run_cmd(f"kubectl get pods --no-headers | grep {service} | head -n 1 | awk '{{print $1\"|\"$3}}'")
        if code != 0 or not stdout:
            return False, None
        
        parts = stdout.split('|')
        pod_name = parts[0]
        status = parts[1]
    else:
        try:
            data = json.loads(stdout)
            if not data.get("items"):
                return False, None
            pod = data["items"][0]
            pod_name = pod["metadata"]["name"]
            status = pod["status"]["phase"]
        except:
            return False, None
            
    report = f"## ☸️ Platform: Kubernetes\n**Service**: {pod_name}\n**Status**: {status}\n\n"
    
    top_out, _, top_code = run_cmd(f"kubectl top pod {pod_name} --no-headers")
    if top_code == 0 and top_out:
        report += f"**Resources (CPU/RAM)**: {top_out}\n\n"
        
    if fetch_logs:
        logs_out, logs_err, logs_code = run_cmd(f"kubectl logs --tail=100 {pod_name}")
        report += "### 📝 Logs (Last 100 lines)\n```text\n"
        combined_logs = (logs_out + "\n" + logs_err).strip()
        report += (combined_logs if combined_logs else "<empty>") + "\n```\n"
        
    return True, report

def main():
    if len(sys.argv) < 2:
        print("Usage: python query_infra.py <target_service> [fetch_logs: true/false]")
        sys.exit(1)
        
    service = sys.argv[1]
    fetch_logs = True if len(sys.argv) > 2 and sys.argv[2].lower() == 'true' else False
    
    print(f"# Infrastructure Report for '{service}'\n")
    
    # Try Docker
    found, report = query_docker(service, fetch_logs)
    if found:
        print(report)
        sys.exit(0)
        
    # Try Kubernetes
    found, report = query_kubernetes(service, fetch_logs)
    if found:
        print(report)
        sys.exit(0)
        
    print(f"[ERROR] Service '{service}' not found in local Docker or Kubernetes context.")
    sys.exit(1)

if __name__ == "__main__":
    main()
